### PR TITLE
Add options for silencing warnings for behaviours

### LINKED ISSUE #8985: An option to silence behaviour-related warnings
**Is your feature request related to a problem? Please describe.**
I want to be able to compile a module like the following one with no warnings:

```erlang
-module(amazing).
-behaviour(gen_server).
```

Currently, if I try to compile it, I get…

```erlang
   ┌─ amazing.erl:
   │
 2 │  -behaviour gen_server.
   │   ╰── Warning: undefined callback function handle_call/3 (behaviour 'gen_server')

   ┌─ amazing.erl:
   │
 2 │  -behaviour gen_server.
   │   ╰── Warning: undefined callback function handle_cast/2 (behaviour 'gen_server')

   ┌─ amazing.erl:
   │
 2 │  -behaviour gen_server.
   │   ╰── Warning: undefined callback function init/1 (behaviour 'gen_server')
```

Which is expected and fine, but in the same way that we have options to disable other warnings (e.g. `nowarn_unused_function`, `nowarn_keywords`, `nowarn_update_literal`, etc.), I would like to have an option to avoid warnings like the ones above.

**Describe the solution you'd like**
I would like the addition of the following compiler options:
* `nowarn_conflicting_behaviours`
* `nowarn_undefined_behaviour_func`
* `nowarn_undefined_behaviour`
* `nowarn_undefined_behaviour_callbacks`
* `nowarn_ill_defined_behaviour_callbacks`
* `nowarn_ill_defined_optional_callbacks`
* `nowarn_behaviour_info`
* `nowarn_redefine_optional_callback`
* `nowarn_undefined_callback`

**Describe alternatives you've considered**
I tried adding `-compile([no_lint]).` to my module, but the warnings were emitted anyway.

**Additional context**
This is the source of the necessity: https://github.com/inaka/elvis_core/pull/367#discussion_r1814971880

### PR DESCRIPTION
**Pull request by bjorng**
This commit adds the following compiler options for suppressing warnings having to do with behaviours:

* nowarn_conflicting_behaviours
* nowarn_undefined_behaviour_func
* nowarn_undefined_behaviour
* nowarn_undefined_behaviour_callbacks
* nowarn_ill_defined_behaviour_callbacks
* nowarn_ill_defined_optional_callbacks

Closes #8985

### CODE REVIEW

**Comment by michalmuskala** [comment]
Should there be an option for completely skipping any behaviour-related checks?

In most cases those checks can be performed by xref or dialyzer without actually loading the module into the VM. Offloading the check to those tools (in projects that use them) could simplify compilation significantly and increase scalability of the compiler by removing compile-time dependencies between modules.

**Comment by bjorng** [comment]
> Should there be an option for completely skipping any behaviour-related checks?

Yes, that makes sense. I've added `nowarn_behaviours` to disable all behaviour-related warnings.

File: lib/stdlib/src/erl_lint.erl [hunk 1]
```diff
@@ -1256,10 +1290,21 @@ all_behaviour_callbacks([{Anno,B}|Bs], Acc, St0) ->
...
 
+add_behaviour_warning(Anno, Warning, St) when is_tuple(Warning) ->
+    Tag = element(1, Warning),
+    case is_warn_enabled(Tag, St) of
+        true ->
+            add_warning(Anno, Warning, St);
+        false ->
+            St
+    end.
+
```
Comment (kikofernandez) [inline lib/stdlib/src/erl_lint.erl:1301]:
```suggestion
add_behaviour_warning(Anno, Warning, St) when is_tuple(Warning) ->
    Tag = element(1, Warning),
    Command = fun () -> add_warning(Anno, Warning, St) end),
    if_warn_enabled(Tag, St, Command.

if_warn_enabled(Tag, St, Command) ->
    case is_warn_enabled(Tag, St) of
        true ->
            Command();
        false ->
            St
    end.
```
This is a suggestion to a common pattern in this PR.
Take it if you find it useful

File: lib/stdlib/src/erl_lint.erl [hunk 2]
```diff
@@ -1231,8 +1260,13 @@ post_traversal_check(Forms, St0) ->
...
 
-check_behaviour(St0) ->
-    behaviour_check(St0#lint.behaviour, St0).
+check_behaviour(St) ->
+    case is_warn_enabled(behaviours, St) of
+        true ->
+            behaviour_check(St#lint.behaviour, St);
+        false ->
+            St
+    end.
```
Comment (kikofernandez) [inline lib/stdlib/src/erl_lint.erl:1269]:
```suggestion
check_behaviour(St) ->
    Command = fun () -> behaviour_check(St#lint.behaviour, St) end,
    if_warn_enabled(behaviours, St, Command).
```
Take this is you like, must be applied with the other suggestion below

**Review by kikofernandez — APPROVED** [review]
Good PR. I just added some suggestions in case they help

File: lib/stdlib/src/erl_lint.erl (hunk 1 above)
Comment (bjorng) [reply lib/stdlib/src/erl_lint.erl:1301]: Thanks for the suggestion. I will save it for a clean-up of `erl_lint` that I am planning to do.

**Pull request merged**